libvast/include/vast/synopsis.hpp

```diff
@@ -115,4 +108,66 @@ pack(flatbuffers::FlatBufferBuilder& builder, const synopsis_ptr&,
 [[nodiscard]] caf::error
 unpack(const fbs::synopsis::LegacySynopsis&, synopsis_ptr&);
 
+/// helper function implemented in cpp as the factory<synopsis> can't be used in
+/// deserialize function below (factory must include synopsis.hpp)
+synopsis_ptr make_synopsis(const type& t);
+
+/// TODO: Serializing and deserializing a synopsis still involves conversion
+/// to/from legacy types. We need to change the synopsis FlatBuffers table to
+/// embed a vast.fbs.Type directly. Ideally we can make the synopsis
+/// memory-mappable just like table slices and types at the same time.
+/// @relates synopsis
+
+/// Loads the contents for this synopsis from `source`.
+template <class Inspector>
+caf::error deserialize(Inspector& source, synopsis_ptr& ptr) {
+  // Read synopsis type.
+  legacy_type t;
+  if (auto err = source(t))
+    return err;
+  // Only nullptr has an none type.
+  if (!t) {
+    ptr.reset();
+    return caf::none;
+  }
+  // Deserialize into a new instance.
+  auto new_ptr = make_synopsis(type::from_legacy_type(t));
+  if (!new_ptr)
+    return ec::invalid_synopsis_type;
+  synopsis::supported_inspectors i{std::ref(source)};
+  if (auto err = new_ptr->inspect(i))
+    return err;
+  // Change `ptr` only after successfully deserializing.
+  using std::swap;
+  swap(ptr, new_ptr);
+  return caf::none;
+}
+
+/// Saves the contents (excluding the layout!) of this slice to `sink`.
+template <class Inspector>
+caf::error serialize(Inspector& sink, synopsis_ptr& ptr) {
+  if (!ptr) {
+    static legacy_type dummy;
+    return sink(dummy);
+  }
+  return caf::error::eval(
+    [&] {
+      return sink(ptr->type().to_legacy_type());
+    },
+    [&] {
+      synopsis::supported_inspectors i{std::ref(sink)};
+      return ptr->inspect(i);
+    });
+}
+
+template <class Inspector>
+  requires(std::is_same_v<typename Inspector::result_type, caf::error>)
```

Comment:
This will be gone in main CAF PR. It is needed as stringification inspector triggers compilation error here.